### PR TITLE
[sparsity] Remove the pack_param from the sparsifier state_dict

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65295
* #65296
* #65294
* #65293
* __->__ #65292

That was the original design, that we decided to simplify by removing the packing in the sparsifier.
The state of the sparsifier is saved directly, and the old behavior accidentally bled through to the current version.
This change removes the `_pack_params` method, and changes the state_dict to include the state directly.
We don't have to change the load_state_dict, as it will work with either the old or the new format.

The main reason for this PR is the simplification. The original design didn't achieve anything useful by packing the sparsification parameters.

Differential Revision: [D31186826](https://our.internmc.facebook.com/intern/diff/D31186826)

### CODE REVIEW

File: torch/ao/sparsity/sparsifier/base_sparsifier.py [hunk 1]
```diff
@@ -183,6 +163,7 @@ def _prepare(self, *args, **kwargs):
             module = config['module']
             param = config.get('parametrization', FakeSparsity)
             mask = config.get('mask', torch.ones(module.weight.shape))
+            self.state[config['fqn']]['mask'] = mask
```
Comment (supriyar) [inline torch/ao/sparsity/sparsifier/base_sparsifier.py:166]: What is 'fqn' here?

Comment (z-a-f) [reply torch/ao/sparsity/sparsifier/base_sparsifier.py:166]: We use 'fqn' for a layer's "fully qualified name". Basically, it's the path that you can refer to using a series of `getattr` statements. I will document it in the docstrings and the design doc

File: torch/ao/sparsity/sparsifier/utils.py [hunk 2]
```diff
@@ -36,3 +36,9 @@ def __init__(self, mask):
     def forward(self, x):
         assert self.mask.shape == x.shape
         return self.mask * x
+
+    def state_dict(self, *args, **kwargs):
+        # We don't want to let the parametrizations to save the mask.
```
Comment (supriyar) [inline torch/ao/sparsity/sparsifier/utils.py:41]: Since we are parametrizing the linear weights, why do we want to not store the mask? Shouldn't this be visible in the linear module state_dict as well?

**Review by supriyar — APPROVED** [review]

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31186826).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31186826).

File: torch/ao/sparsity/sparsifier/utils.py (hunk 2 above)
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier/utils.py:41]: That was one of the design choices that we trhough made sense -- the linear layer uses parameterizations, which are created (and managed) by the sparsifier. When the sparsifier is detached (squashed_mask) the parameterization and the mask is removed from the linear layer.

Given the above, if store the linear layer using the `state_dict` it will store the parametrizations as well. However, the ownership belongs to Sparsifier (by our definitions), which also stores the mask when `state_dict` is dumped. To avoid duplication, we decided to not let linear layer serioalize the fakesparse parameterization.
